### PR TITLE
[KEYCLOAK-2343] - Allow exact user search by user attributes

### PR DESCRIPTION
**Pull request by pedroigor**

### CODE REVIEW

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/ldap/LDAPGroupMapperTest.java [hunk 1]
```diff
@@ -74,6 +74,11 @@ protected void afterImportTestRealm() {
     @Test
     public void test01_ldapOnlyGroupMappings() {
         test01_ldapOnlyGroupMappings(true);
+        try {
+            Thread.sleep(1000000000);
```
Comment (stianst) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/ldap/LDAPGroupMapperTest.java]: ?????

Comment (pedroigor) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/ldap/LDAPGroupMapperTest.java]: Oops !

**Comment by Hokwang** [comment]
please think release plan this issue once again, I think you can push to 9.0.x not 10.0.0.
everyone want to this and this is not big thing, isn't it?

**Review by hmlnarik — COMMENTED** [review]
Looks good to me, @pedroigor , thanks! Few minor things and it's good to merge.

File: integration/admin-client/src/main/java/org/keycloak/admin/client/resource/UsersResource.java [hunk 2]
```diff
@@ -56,6 +56,10 @@
     @Produces(MediaType.APPLICATION_JSON)
     List<UserRepresentation> search(@QueryParam("username") String username);
 
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    List<UserRepresentation> search(@QueryParam("username") String username, @QueryParam("exact") Boolean exact);
```
[the remark this answers is not in the document]
Comment (pedroigor) [reply integration/admin-client/src/main/java/org/keycloak/admin/client/resource/UsersResource.java:61]: I see your point. However, that is going to break people using that method. What I can do is add a new variant of it with the new parameter.

The method you mentioned means also another type of query that looks up multiple properties (iirc, email, username, first/last names).

This one is specific for username and for other usages the API is ready. The focus here is really the API.

**Review by hmlnarik — DISMISSED** [review]
LGTM, thanks!

**Review by hmlnarik — APPROVED** [review]

**Comment by fzyzcjy** [comment]
Hi this sound very helpful! I wonder -

1. When will this become released? (seems not in 10.0
2. Can we search by exact email when using Admin Client java code? (I only [see](https://github.com/keycloak/keycloak/pull/6926/files#diff-714819e3686a2b8815d5a4997fca372bR60) the "exact username"...)

**Pull request merged**